### PR TITLE
[devp2p discovery]: cleanup

### PR DESCRIPTION
**Pull request by niklasad1**
Fixes:
- Some formatting
- Replace manual default implementation
- Lifetime
- Packet_ID api, just pass by value and avoid cloning internally.

It can be merged after #11540 to avoid conflicts.

### CODE REVIEW

File: util/network-devp2p/src/discovery.rs [hunk 1]
```diff
@@ -191,7 +183,7 @@ pub struct TableUpdates {
 }
 
 impl<'a> Discovery<'a> {
-	pub fn new(key: &KeyPair, public: NodeEndpoint, ip_filter: IpFilter) -> Discovery<'static> {
+	pub fn new(key: &KeyPair, public: NodeEndpoint, ip_filter: IpFilter) -> Discovery<'a> {
```
Comment (ordian) [inline util/network-devp2p/src/discovery.rs]: it seems the previous `'static` lifetime was valid, why the change?

Comment (ordian) [reply util/network-devp2p/src/discovery.rs]: We can removed the lifetime altogether by specifying `request_backoff: &'static [Duration]` in the struct

Comment (niklasad1) [reply util/network-devp2p/src/discovery.rs]: > it seems the previous 'static lifetime was valid, why the change?

I thought the `static lifetime` for `request_backoff: &'a [Duration]` unnecessarily big? That's why, i.e, needs only `live` as long the `"a` AFAIU.

It seems like it's never changed after construction, so could could be `&'static Duration` I guess.

Any strong opinions?

Comment (ordian) [reply util/network-devp2p/src/discovery.rs]: I think the reason the struct has a lifetime in the first place is to be able to pass any reference for `request_backoff`, but in practice it's always a (non-mut) reference to a static var, so I think we should remove the lifetime altogether.

File: util/network-devp2p/src/discovery.rs [hunk 2]
```diff
@@ -389,14 +382,14 @@ impl<'a> Discovery<'a> {
 		node.endpoint.to_rlp_list(&mut rlp);
 		append_expiration(&mut rlp);
 		let old_parity_hash = keccak(rlp.as_raw());
-		let hash = self.send_packet(PACKET_PING, &node.endpoint.udp_address(), &rlp.drain())?;
+		let hash = self.send_packet(PACKET_PING, node.endpoint.udp_address(), &rlp.drain())?;
```
Comment (dvdplm) [inline util/network-devp2p/src/discovery.rs]: Do you think it would make sense to move `rlp` to `send_packet()` to avoid the copying in `assemble_packet()`?

File: util/network-devp2p/src/discovery.rs [hunk 3]
```diff
@@ -389,14 +382,14 @@ impl<'a> Discovery<'a> {
...
 
 		self.in_flight_pings.insert(node.id, PingRequest {
 			sent_at: Instant::now(),
 			node: node.clone(),
 			echo_hash: hash,
 			deprecated_echo_hash: old_parity_hash,
```
Comment (dvdplm) [inline util/network-devp2p/src/discovery.rs]: Is there a chance we can get rid of this you think?

File: util/network-devp2p/src/discovery.rs [hunk 4]
```diff
@@ -419,10 +412,10 @@ impl<'a> Discovery<'a> {
 		Ok(())
 	}
 
-	fn send_packet(&mut self, packet_id: u8, address: &SocketAddr, payload: &[u8]) -> Result<H256, Error> {
+	fn send_packet(&mut self, packet_id: u8, address: SocketAddr, payload: &[u8]) -> Result<H256, Error> {
 		let packet = assemble_packet(packet_id, payload, &self.secret)?;
 		let hash = H256::from_slice(&packet[0..32]);
```
Comment (dvdplm) [inline util/network-devp2p/src/discovery.rs]: This is a bit annoying, we just built one of these in `assemble_packet()`; maybe return the `signed_hash` from there?

Comment (niklasad1) [reply util/network-devp2p/src/discovery.rs]: hehe, right I do get your concern

In terms, of `clean code` if we change `assemble_packet` then it does clearly more than one thing but it could be quite nice if we could introduce another type `Packet (or similar)` to avoid `copy_from_slice` stuff that would quite nice.

But, let's address it another PR :)

File: util/network-devp2p/src/discovery.rs (hunk 2 above)
Comment (niklasad1) [reply util/network-devp2p/src/discovery.rs]: Yeah, great catch

File: util/network-devp2p/src/discovery.rs (hunk 3 above)
Comment (niklasad1) [reply util/network-devp2p/src/discovery.rs]: what do you mean?

Comment (dvdplm) [reply util/network-devp2p/src/discovery.rs]: No, I mean `deprecated_echo_hash` – I guess we must have made a change here at some point and kept the old hash around for backwards compatibility. Do we still have to include it you think? Can we ever remove it?

File: util/network-devp2p/src/discovery.rs (hunk 4 above)
Comment (dvdplm) [reply util/network-devp2p/src/discovery.rs]: Do you really think having `assemble_packet` return `Result<(Bytes, H256), Error>` would be so bad? It's a private method after all, and it wouldn't be doing more or different work?

> But, let's address it another PR :)

Absolutely.

Comment (niklasad1) [reply util/network-devp2p/src/discovery.rs]: >Do you really think having assemble_packet return Result<(Bytes, H256), Error> would be so bad? It's a private method after all, and it wouldn't be doing more or different work?

It is fine but my general opinion it should not :)

File: util/network-devp2p/src/discovery.rs (hunk 3 above)
Comment (niklasad1) [reply util/network-devp2p/src/discovery.rs]: Probably but I'm not sure.
